docs/Guidelines.md

```diff
@@ -0,0 +1,78 @@
+  example: `KLOG_VERBOSITY=9 ./mage runE2ETests` would output `curl` commands
+  issued via Kubernetes client from sigs.k8s.io/controller-runtime
+* To quickly debug a test, you can run only the desired suite. Example: `./bin/e2e-appstudio --ginkgo.focus="e2e-demos-suite"`
+* Split tests in different scenarios. Will be better to debug a small scenario than a very big one
+
+## Debuggability
+
+If your test fails, it should provide as detailed as possible reasons for the failure in its failure message. The failure message is the string that gets passed (directly or indirectly) to ginkgo.Fail[f].
+
+A good failure message:
+* identifies the test failure
+* has enough details to provide some initial understanding of what went wrong
+
+It is good practice to include details like the object that failed some assertion in the failure message because then a) the information is available when analyzing a failure that occurred in the CI and b) it only gets logged when some assertion fails. Always dumping objects via log messages can make the test output very large and may distract from the relevant information.
+
+Dumping structs with `format.Object` is recommended. Starting with Kubernetes 1.26, format.Object will pretty-print Kubernetes API objects or structs as YAML and omit unset fields, which is more readable than other alternatives like `fmt.Sprintf("%+v")`.
+
+```golang
+import (
+    "fmt"
+    "k8s.io/api/core/v1"
+    "k8s.io/kubernetes/test/utils/format"
+)
+
+var pod v1.Pod
+fmt.Printf("format.Object:\n%s", format.Object(pod, 1 /* indent one level */))
+
+format.Object:
+    <v1.Pod>:
+        metadata:
+          creationTimestamp: null
+        spec:
+          containers: null
+        status: {}
+```
+## Polling and timeouts
+
+When waiting for something to happen, use a reasonable timeout. Without it, a test might keep running until the entire test suite gets killed by the CI. **Beware that the CI under load may take a lot longer to complete some operation compared to running the same test locally**. On the other hand, a too long timeout also has drawbacks:
+
+* When a feature is broken so that the expected state doesn’t get reached, a test waiting for that state first needs to time out before the test fails.
+* If a state is expected to be reached within a certain time frame, then a timeout that is much higher will cause test runs to be considered successful although the feature was too slow. A dedicated performance test in a well-know environment may be a better solution for testing such performance expectations.
+
+Good code that waits for something to happen meets the following criteria:
+* accepts a context for test timeouts
+* full explanation when it fails: when it observes some state and then encounters errors reading the state, then dumping both the latest observed state and the latest error is useful
+* early abort when condition cannot be reached anymore
+
+### Tips for writing and debugging long-running tests
```

Comment:
Do we want to make reference and recommend that they generate their tests using our tooling documented in the DeveloperGeneratingTest.doc?

Reply:
We already do that in the main page. Not sure if is ok to repeat this so much time